### DISCUSSION TITLE
Is there anyway to compute how much memory will be used in GPU?

### DISCUSSION THREAD

**Discussion by jiecaix** [BLAST-WarpX/warpx#5072] [Q&A] [answered]
I want to calculate the memory used in GPU to init a memory pool, is there anyway to do it. Assume the uniform particle density and the fix size of grid.

**Comment by ax3l** [comment, marked as answer]
Hi @zhazhajust,

While this depends on the exact simulation parameters, you can estimate the basic parameters by using 9+ (x,y,z,w,px,py,pz,cpuid, ...) doubles (8 byte each) per particle, and 9+ doubles (E,B,j, ...) per cell for the basic EM PIC algorithm.

Then, any additional attributes (e.g., ionization, QED) or algorithms (e.g., PSATD) add to those.

We then need some temporary memory, e.g., to sort particles or communicate them (select, copy, put back in data structures on a different node), etc. Depending on details, that is 1.5-2.5x overhead.

@n01r is trying to give a basic estimate with his calculator in
https://github.com/ECP-WarpX/WarpX/pull/3296
as well, which we try to merge to allow simpler estimates.
